### PR TITLE
Allow items to use multiple ammotypes

### PR DESCRIPTION
**Pull request by ymber**
#### Summary
```SUMMARY: Features "Allow items to use multiple ammotypes"```

#### Purpose of change
Defining an array of allowed ammotypes in a gun definition is much better than the hack some mods currently use where an ammo item pretends to be several different rounds. Magazines can also take different rounds eg loading .458 SOCOM in a STANAG magazine.

#### Describe the solution
Use a std::set<ammotype> in items, update ammo_type to return this set, and update the rest of the code to handle the set properly. Even though magazines can take different ammotypes now I left the std::map<ammotype, itype_id> for default magazines to make it easier to pick a default magazine for each ammotype.

### CODE REVIEW

**Comment by tenmillimaster** [comment]
Yay! The Lw12 magazines can get fixed when done!

File: src/character.cpp [hunk 1]
```diff
@@ -1009,13 +1009,18 @@ void find_ammo_helper( T &src, const item &obj, bool empty, Output out, bool nes
             }
             if( node->is_ammo_container() && !node->contents.empty() &&
                 !node->contents_made_of( SOLID ) ) {
-                if( node->contents.front().type->ammo->type.count( ammo ) ) {
-                    out = item_location( src, node );
+                for( ammotype at : ammo ) {
```
Comment (ifreund) [inline src/character.cpp]:
```suggestion
                for( const ammotype &at : ammo ) {
```

File: src/item.cpp [hunk 2]
```diff
@@ -5834,12 +5834,12 @@ itype_id item::ammo_current() const
     return ammo ? ammo->get_id() : "null";
 }
 
-ammotype item::ammo_type( bool conversion ) const
+std::set<ammotype> item::ammo_type( bool conversion ) const
```
Comment (ifreund) [inline src/item.cpp]: This function probably needs to be renamed to `ammo_types()` now

**Comment by Night-Pryanik** [comment]
Yay! Pneumatic weapons with compressed air AND BBs, at last!

File: src/character.cpp [hunk 3]
```diff
@@ -1009,13 +1009,18 @@ void find_ammo_helper( T &src, const item &obj, bool empty, Output out, bool nes
...
+                    if( node->contents.front().type->ammo->type.count( at ) ) {
+                        out = item_location( src, node );
+                    }
                 }
                 return VisitResponse::SKIP;
             }
-            if( node->is_ammo() && node->type->ammo->type.count( ammo ) ) {
-                out = item_location( src, node );
+
+            for( ammotype at : ammo ) {
```
Comment (ifreund) [inline src/character.cpp]:
```suggestion
            for( const ammotype &at : ammo ) {
```

File: src/dump.cpp [hunk 4]
```diff
@@ -187,7 +187,9 @@ bool game::dump_stats( const std::string &what, dump_mode mode,
         auto dump = [&rows, &locations]( const standard_npc & who, const item & obj ) {
             std::vector<std::string> r;
             r.push_back( obj.tname( 1, false ) );
-            r.push_back( obj.ammo_type() ? obj.ammo_type().str() : "" );
+            for( ammotype at : obj.ammo_type() ) {
```
Comment (ifreund) [inline src/dump.cpp]:
```suggestion
            for( const ammotype &at : obj.ammo_type() ) {
```

File: src/item.h [hunk 5]
```diff
@@ -1585,7 +1585,7 @@ class item : public visitable<item>
         /** Ammo type (@ref ammunition_type) used by item
          *  @param conversion whether to include the effect of any flags or mods which convert the type
          *  @return NULL if item does not use a specific ammo type (and is consequently not reloadable) */
-        ammotype ammo_type( bool conversion = true ) const;
+        std::set<ammotype> ammo_type( bool conversion = true ) const;
```
Comment (ifreund) [inline src/item.h]: Need to update the above comment.

**Comment by ymber** [comment]
@ifreund I want to avoid ammo types that are lists of other ammo types. Defining those can get complex while `"ammo": ["357mag", "38"]` does the same thing with simpler data.

**Comment by YukariMindGap** [comment]
YEs! I can make weapons that use both battery and ammo!

**Comment by tenmillimaster** [comment]
@ifreund the other problem is that it doesn't do justice to the use case in the PR (e.g. 458 socom in a STANAG.) It worked well for ammotypes that had inheritance, like .38 and .357, but situations where they don't (any of the numerous ammos that can share a magazine but cannot be fired out of the same gun) it would produce undesired results, or require silly work-arounds. That being creating a *3rd* ammotype that's a parent of the other two, just for the magazines, and then quieting the  json parser on the 'magazine mismatch' error.

**Comment by ymber** [comment]
I reckon this is in a fit state to start testing now. There's still some stuff needs working on but it's most of the way there.
- [ ] Look at what new json validation is needed
- [ ] Look at what new unit tests should be implemented
- [ ] Update documentation for affected json

For future work:
- [ ] Define magazine capacity in `[ "ammotype", (int)capacity ]` pairs
- [ ] Allow loading multiple ammotypes at once from multiple magazines eg for pneumatics

**Comment by tinukedaya** [comment]
So, I wonder, would this be possible now? #30319?

**Comment by ymber** [comment]
Closer to possible than it was but you'd still need UI work for it and some fairly significant infrastructure.

File: src/dump.cpp [hunk 6]
```diff
@@ -187,7 +185,10 @@ bool game::dump_stats( const std::string &what, dump_mode mode,
         auto dump = [&rows, &locations]( const standard_npc & who, const item & obj ) {
             std::vector<std::string> r;
             r.push_back( obj.tname( 1, false ) );
-            r.push_back( obj.ammo_type() ? obj.ammo_type().str() : "" );
+            const std::set<ammotype> atypes = obj.ammo_types();
+            r.push_back( !atypes.empty() ? enumerate_as_string( atypes.begin(), atypes.end(), []( const ammotype &at ) {
```
Comment (kevingranade) [inline src/dump.cpp]:
```suggestion
            r.push_back( !atypes.empty() ? enumerate_as_string( atypes.begin(),
            atypes.end(), []( const ammotype &at ) {
```

File: src/item.cpp [hunk 7]
```diff
@@ -6432,7 +6502,7 @@ bool item::reload( player &u, item_location loc, int qty )
             qty = std::min( qty, ammo->ammo_remaining() );
             ammo->ammo_consume( qty, tripoint_zero );
             charges += qty;
-        } else if( ammo_type() == ammotype( "plutonium" ) ) {
+        } else if( ammo->type->ammo->type == "plutonium" ) {
```
Comment (ifreund) [inline src/item.cpp]: Looks like you are checking this same `ammo->type->ammo->type` thing quite often, and sometimes without checking to ensure `ammo->type->ammo` isn't a `cata::nullopt`. 

I'd recommend making a new function in item that checks if the item is ammo and then preforms this comparison.

File: src/item.cpp [hunk 8]
```diff
@@ -5861,12 +5888,12 @@ itype_id item::ammo_current() const
     return ammo ? ammo->get_id() : "null";
 }
 
-ammotype item::ammo_type( bool conversion ) const
+std::set<ammotype> item::ammo_types( bool conversion ) const
```
[the remark this answers is not in the document]
Comment (ymber) [reply src/item.cpp]: See #31393.

**Comment by IvanShestakov** [comment]
@ymber can this be the cause for #31398 ?

**Comment by ymber** [comment]
Possibly. I'll look into it.

**Pull request merged**